src/managers/KeybindManager.cpp

```diff
@@ -2074,15 +2105,65 @@ void CKeybindManager::moveOutOfGroup(std::string args) {
     if (!PWINDOW || !PWINDOW->m_sGroupData.pNextWindow)
         return;
 
-    g_pLayoutManager->getCurrentLayout()->onWindowRemoved(PWINDOW);
+    moveWindowOutOfGroup(PWINDOW);
+}
 
-    const auto GROUPSLOCKEDPREV = g_pKeybindManager->m_bGroupsLocked;
+void CKeybindManager::moveWindowOrGroup(std::string args) {
+    char               arg = args[0];
 
-    g_pKeybindManager->m_bGroupsLocked = true;
+    static auto* const BCHECKGROUPLOCK = &g_pConfigManager->getConfigValuePtr("binds:check_group_lock")->intValue;
 
-    g_pLayoutManager->getCurrentLayout()->onWindowCreated(PWINDOW);
+    if (!isDirection(args)) {
+        Debug::log(ERR, "Cannot move into group in direction %c, unsupported direction. Supported: l,r,u/t,d/b", arg);
+        return;
+    }
 
-    g_pKeybindManager->m_bGroupsLocked = GROUPSLOCKEDPREV;
+    const auto PWINDOW = g_pCompositor->m_pLastWindow;
+    if (!PWINDOW)
+        return;
+    const auto ISWINDOWGROUP       = PWINDOW->m_sGroupData.pNextWindow;
+    const auto ISWINDOWGROUPLOCKED = ISWINDOWGROUP && PWINDOW->getGroupHead()->m_sGroupData.locked;
+
+    const auto PWINDOWINDIR             = g_pCompositor->getWindowInDirection(PWINDOW, arg);
+    const auto ISWINDOWINDIRGROUP       = PWINDOWINDIR && PWINDOWINDIR->m_sGroupData.pNextWindow;
+    const auto ISWINDOWINDIRGROUPLOCKED = ISWINDOWINDIRGROUP && PWINDOWINDIR->getGroupHead()->m_sGroupData.locked;
+
+    // note: PWINDOWINDIR is not null implies !PWINDOW->m_bIsFloating
+    if (ISWINDOWINDIRGROUP && !ISWINDOWINDIRGROUPLOCKED) {
+        if (ISWINDOWGROUPLOCKED && *BCHECKGROUPLOCK) {
+            g_pLayoutManager->getCurrentLayout()->switchWindows(PWINDOW, PWINDOWINDIR);
+            g_pCompositor->warpCursorTo(PWINDOW->middle());
+        } else {
+            moveWindowIntoGroup(PWINDOW, PWINDOWINDIR);
+        }
+    } else if (ISWINDOWINDIRGROUPLOCKED) {
+        if (*BCHECKGROUPLOCK) {
+            g_pLayoutManager->getCurrentLayout()->switchWindows(PWINDOW, PWINDOWINDIR);
+            g_pCompositor->warpCursorTo(PWINDOW->middle());
+        } else {
+            moveWindowIntoGroup(PWINDOW, PWINDOWINDIR);
+        }
+    } else if (PWINDOWINDIR) {
+        if (ISWINDOWGROUP && (!*BCHECKGROUPLOCK || !ISWINDOWGROUPLOCKED)) {
+            moveWindowOutOfGroup(PWINDOW);
+        } else {
+            g_pLayoutManager->getCurrentLayout()->switchWindows(PWINDOW, PWINDOWINDIR);
+            g_pCompositor->warpCursorTo(PWINDOW->middle());
+        }
+    } else {
+        if (ISWINDOWGROUP && (!*BCHECKGROUPLOCK || !ISWINDOWGROUPLOCKED)) {
+            moveWindowOutOfGroup(PWINDOW);
+        }
+    }
+}
+
+void CKeybindManager::setGroupLockChecking(std::string args) {
+    static auto* const BCHECKGROUPLOCK = &g_pConfigManager->getConfigValuePtr("binds:check_group_lock")->intValue;
+
+    if (args == "toggle")
```

Comment:
I was thinking of using a different variable at runtime, e.g. `bCheckGroupLock`, and just using `binds:check_group_lock` as the initial value, so that reloading the configuration at runtime won't have any effect as the dispachter will only check for `bCheckGroupLock`.

However this adds complexity and probably betrays the purpose of making it a configuration variable, regardless of who takes precedence.

Reply:
Also, what do you think about adding some sort of indication that the check group lock is enabled, such as in #2507, or posting an event?

Reply:
I'm going to leave here now, as this dispatcher is not meant to permanently change the lock-checking behaviour.